statsd/receiver.go

```diff
@@ -149,24 +151,16 @@ func (mr *MetricReceiver) handleMessage(addr net.Addr, msg []byte) {
 		if len(line) > 1 {
 			metric, err := mr.parseLine(line)
 			if err != nil {
-				log.Errorf("Error parsing line %q from %s: %v", line, addr, err)
+				log.Warnf("Error parsing line %q from %s: %v", line, addr, err)
 				mr.increment("bad_lines_seen", 1)
 				continue
 			}
-			source := strings.Split(addr.String(), ":")
-			hostname := source[0]
-			if net.ParseIP(hostname) != nil {
-				if mr.Cloud != nil {
-					instance, err := cloudprovider.GetInstance(mr.Cloud, hostname)
-					if err != nil {
-						log.Errorf("Error retrieving instance details from cloud provider %s: %v", mr.Cloud.ProviderName(), err)
-					} else {
-						hostname = instance.ID
-						metric.Tags = append(metric.Tags, instance.Tags...)
-						metric.Tags = append(metric.Tags, fmt.Sprintf("%s:%s", "region", instance.Region))
-					}
-				}
-				metric.Tags = append(metric.Tags, fmt.Sprintf("%s:%s", types.StatsdSourceID, hostname))
+			if !triedToGetTags {
```

Comment:
So I thought about this more, and the reason I feel this is unnecessary is that in reality the loop is not likely to run more than once as clients usually send each line independently rather than batching multiple lines in one packet. Any additional state and logic can hide bugs and make it harder to reason about what the program does. And this really is a micro-optimisation. That said, I like that the logic to retrieve tags is extracted to a different function as this is much cleaner so I'm OK to go with your proposition but for two things and a minor issue:

1. Rename `tryGetTags` to `tryToGetTags` (personally I'd probably just keep `getTags` or `getAdditionalTags`)
1. Maybe returns multiple values from `tryToGetTags` i.e. `additionalTags, triedToGetTags = mr.tryGetTags(addr.String())`
1. `tryToGetTags` should return an empty slice, not `nil` i.e. `return types.Tags{}`